src/Xamarin.Android.Build.Tasks/Utilities/LlvmIrGenerator/StringHolder.cs

```diff
@@ -0,0 +1,124 @@
+using System;
+
+namespace Xamarin.Android.Tasks.LLVMIR;
+
+class StringHolder : IComparable, IComparable<StringHolder>, IEquatable<StringHolder>
+{
+	public LlvmIrStringEncoding Encoding { get; }
+	public string? Data { get; }
+
+	StringComparison comparison;
+
+	public StringHolder (string? data, LlvmIrStringEncoding encoding = LlvmIrStringEncoding.UTF8, StringComparison comparison = StringComparison.Ordinal)
+	{
+		Data = data;
+		Encoding = encoding;
+		this.comparison = comparison;
+	}
+
+	public static StringHolder AsHolder (object? value, LlvmIrStringEncoding encoding = LlvmIrStringEncoding.UTF8, StringComparison comparison = StringComparison.Ordinal)
+	{
+		if (value == null) {
+			return new StringHolder ((string)value);
+		}
+
+		StringHolder holder;
+		if (value is string) {
+			holder = new StringHolder ((string)value, encoding, comparison);
+		} else if (value is StringHolder) {
+			holder = (StringHolder)value;
+		} else {
+			throw new InvalidOperationException ($"Internal error: expected 'string' type, got '{value.GetType ()}' instead.");
+		}
+
+		return holder;
+	}
+
+	public int CompareTo (object obj)
```

Comment:
Given that `CompareTo(StringHolder?)` already has a null check, this could be simplified to:

```csharp
public int CompareTo (object obj) => CompareTo (obj as StringHolder);
```